### ISSUE TITLE
Use arcgislayers as backend for ESRI's AGOL

### ISSUE THREAD

**Issue by kbvernon** [ropensci/FedData#109]
Hi Kyle, I think I mentioned this to you before, but I want to suggest adopting `{arcgislayers}` as a backend for interacting with ESRI's AGOL. That package is going through some additional testing, but I believe it should reach a stable version in the next couple of weeks or month.

**Benefits:**
1. `{arcgislayers}` is officially supported by ESRI. That's maybe not a great argument... but!
2. It would remove a ton of code from `{FedData}` that you would no longer have to maintain. 
3. It would make it easier to introduce some functionality that we've talked about before, like query support (https://github.com/ropensci/FedData/issues/68) and attribute/column selection.
5. And, it can be done without introducing breaking changes.

**Costs:**
1. Maybe AWS is better? (I have almost no experience with AWS and am not sure this is entirely true, so take it for what it is.)
2. `{arcgislayers}` is built around the idea of _interacting_ with AGOL (upload, download, query, retrieve metadata, etc), so there's a risk of mission creep. 

Just to give an example of the trade-off here, `{arcgislayers}` gives you the power to query AGOL and return a table without spatial data, which is nice for exploring data before figuring out what features to download. It also allows you to specify what attribute data you want to get. But right now, `{FedData}` always returns all of the spatial and attribute data. My own feeling is that the benefits of filtering rows and selecting columns outweigh the potential risks, but a hardline should definitely be drawn there. 

Also, `get_nhd()` will require some special treatment since it returns a list. My guess is additional functions for retrieving individual components of NHD, so that uber function can be left alone.

What do you think? 

If it helps, this is something I would be more than willing to invest some time. Least I can do since this package factors so heavily in all of my research.

**Comment by bocinsky** [comment]
I completely agree — the benefits outweigh the costs here. If you are willing to prototype something and submit a PR (or even just fork and share your repo with me), I think we can get this done. I could use the help!

My goals would be to add functionality to the api but not create any breaking changes at this point. We can talk about restructuring what functions return (e.g., `get_nhd()`) at a later juncture.

Thanks much for the offer!

Best,
Kyle

**Comment by kbvernon** [comment]
Thanks, Kyle! I'm going to wait till they release a stable version before I start working on this (probably mid-February), then I can share a PR.

Cheers.

**Comment by bocinsky** [comment]
Sounds good. Thank you!

**Comment by JosiahParry** [comment]
If you all end up taking this on, please let me know if I can be of assistance! I'm sure y'all will run into some "nice to haves" and if you do, please let me know. Additionally, if there is any progress made on this before March 10th-ish, I'd love to talk about it at Esri DevSummit

**Comment by bocinsky** [comment]
Will do. Thanks @JosiahParry!

**Comment by bocinsky** [comment]
Hi y'all. I've begun the implementation of this, first by simply using `arcgislayers` without changes to the `FedData` api. So, query support requested in (https://github.com/ropensci/FedData/issues/68) is not yet implemented. But it works well, and I'm pleased to rely on an official ESRI package for accessing ArcGIS layers. 

That being said, **we can't release this update to CRAN until `arcgislayers` is on CRAN**.  I didn't put two and two together about this until I'd already implemented it. @JosiahParry, any notion on if/when a CRAN release of `arcgislayers` is likely to happen?

**Comment by JosiahParry** [comment]
Yup! Totally fair point. :))) 
@bocinsky The goal is a CRAN release within the next few weeks. Before March 12! I need to sniff out as many bugs as possible and fix them before then. So...if you have any that you've encountered please do let me know. I'll keep you posted on this. Sorry for the headache!

**Comment by bocinsky** [comment]
It's all good! No bugs so far, and thanks for handling those messages for zero feature results!

**Comment by JosiahParry** [comment]
arcgisutils 0.2.0 is released on CRAN 

arcgislayers is waiting for review: https://cran.r-project.org/incoming/newbies/

**Comment by bocinsky** [comment]
Hey, congratulations! 🎉 I'll switch FedData to use the CRAN version. Cheers!

**Comment by kevinwolz** [comment]
@bocinsky I have run into an issue when trying to use `FedData >=4.0.1` within a `rocker/geospatial:4.4.2` Docker container (the latest geospatial docker container of this kind). Installing `FedData >=4.0.0` works fine, but anything greater than `4.0.1` fails. The fails refer to `arcgisutils`, `arcgislayers`, and `arcpbf` - dependencies which are new starting with `FedData >=4.0.1`.

It seems that this ultimate reason for this is that `arcgisutils` requires `rustc >= 1.67`. The installation was failing saying that `rustc` was not available. I tried adding the following line to the Docker container to install `rustc`,  but it still doesn't install.

```
RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y
```

`FedData >=4.0.1` does install fine on the previous `rocker/geospatial:4.3.3` version, which is what I will have to use for now - I haven't looked deeply into why it installs here and not on the latest `rocker/geospatial`, but it does. I just wanted to let you know that something weird is going on here and should be investigated, as this may make `FedData` difficult to work with as `rocker/geospatial` advances.

**Comment by JosiahParry** [comment]
This is unrelated to this issue. You do not need to have rust installed to install the packages. What you want is a pre built binary for your distribution. Consider using posit public package manager instead of whichever cran mirror you’re using presently

**Comment by kevinwolz** [comment]
For the record, I really don't know much about `rocker/geospatial`, `rust`, or any of these arc packages. So it's probably worth your own investigation if you haven't already. Something just smelled off to me since I know `rocker/geospatial` is very commonly used.

**Comment by JosiahParry** [comment]
@kevinwolz what you're experiencing is a challenge with _compiled_ code. It just so happens that it is with the R package arcgisutils. You may run into similar challenges for which there isn't a pre-build binary particularly if they have specific system requirements. 

For background, the package arcgislayers is an Esri-built R package to access data hosted in ArcGIS Online / Enterprise. The R packages are anywhere from 10-100x faster than the other alternatives. This is because the package uses compiled code. 

R packages that are written entirely in R do not need to be compiled. Packages that use C, C++, Rust or other compiled languages need to be compiled. CRAN will compile these R packages into what are called "binaries" for Mac and Windows. Package binaries are _pre-compiled_ meaning you don't need to have C, C++, or Rust to install the package. Using package binaries is wayyyyy faster to install than compiling the package yourself. 

The issue you're facing is that `rocker/geospatial:latest` uses Ubuntu 24 which is a fairly new release (April, 2024). CRAN _does not_ build binaries for linux distributions. The rocker organization uses [Posit's Public Package Manager (P3M)](https://p3m.dev/) to install pre-compiled R package binaries. P3M provides binaries for `arcgisutils` and `arcpbf` for Ubuntu 22 and RHEL 8, for example. 

The `r-arcgis` R-universe provides Ubuntu 24 binaries for these packages. Below is a repro as to how to install the packages: 

```dockerfile
FROM rocker/r-ver:4.4.2

RUN R -e 'install.packages(c("arcgisutils", "arcpbf"), repos = c("https://r-arcgis.r-universe.dev/bin/linux/noble/4.4", options("repos")))'

ENTRYPOINT [ "R" ]
```

**Comment by zross** [comment]
@JosiahParry I just wanted to thank you for that useful/thoughtful response and where/how to install the binaries for `arcgisutils` and `arcbf`!

**Comment by bocinsky** [comment]
Closing this. I'd switched to `arcgislayers`, but then switched back to remove the extra dependency. We just don't need all the functionality of `arcgislayers`... the API is fairly straightforward.

**Comment by kevinwolz** [comment]
@bocinsky I just noticed that `get_wbd()`still depends on `arcgislayers`. Is that intentional, or can this dependency be removed like you did elsewhere?
